### PR TITLE
pm: WIP for power management that uses submodules

### PR DESCRIPTION
**Pull request by roberthartung**
This commit resolves issues with the power management and prepares the PM architecture to be selected by the actual CPU.

### CODE REVIEW

**Comment by kaspar030** [comment]
It seems to me that you didn't reduce the number of duplicate implementations. How about using submodules in the pm_fallback module, which can be independently selected?

**Comment by roberthartung** [comment]
Hi @kaspar030 , the idea was not to reduce duplication, but to resolve some issues with the current selection of pm functions/implementations and to allow the actual cpu to use the implementation it needs.

Submodules sound the way to go for me! Which would you suggest? One for each of pm_off, pm_reboot and pm_set_lowest?

**Comment by kaspar030** [comment]
> Submodules sound the way to go for me! Which would you suggest? One for each of pm_off, pm_reboot and pm_set_lowest?

Yes! Can we apply this to the cpu fallbacks, too?

**Comment by roberthartung** [comment]
@kaspar030 check out my updated comment!

**Comment by roberthartung** [comment]
@kaspar030 That's the question how it would work with the most used CPUs (atmega_common, stm32_common, cortexm_common and nrf5x_common). Let's try to figure it out from my proposal.

**Comment by kaspar030** [comment]
> Like this? That also looks like a lot of overhead?

Hm, it does, but we do want this fine-grained selection. The alternatives are ifdefs with global defines, weak defaults, ...

Let's try!

**Comment by roberthartung** [comment]
Yes, overhead is relative though, because we can just select fine-grained implementations. The main problem is that some CPUs use both stm32_common AND cortexm_common. And we have to review the current pm implementation and see if we can merge it to one of them IMO.

I would really like to avoid any ifdefs/weak defaults if possible.

**Comment by roberthartung** [comment]
@kaspar030 Can you help me with enabling a simple submodule? What do I need? Let's say I want to change pm_fallback to have three submodules: pm_fallback_off, pm_fallback_set_lowest and pm_fallback_reboot. Do I need to cherry-pick / get anything from #7241 ?

**Comment by kaspar030** [comment]
> @kaspar030 Can you help me with enabling a simple submodule? What do I need? Let's say I want to change pm_fallback to have three submodules: pm_fallback_off, pm_fallback_set_lowest and pm_fallback_reboot. 

The steps should be

- add ```SUBMODULES := 1``` to the pm_fallback Makefile
- move the implementations to "off.c", "set_lowest.c", "reboot.c"
- add "pm_fallback_%" to makefiles/pseudomodules.mk (or pm_fallback% if there are no always compiled .c files).

See how it is done in core/Makefile.

> Do I need to cherry-pick / get anything from #7241 ?

That should be independent. I'll be rebasing it on this one once we're done here. ;)

**Comment by roberthartung** [comment]
@kaspar030 Okay, had a little typo :/

 Only problem now is that I have to add `USEMODULE += pm_fallback` to be able to do the following:
```
USEMODULE += pm_fallback_off
USEMODULE += pm_fallback_set_lowest
USEMODULE += pm_fallback_reboot
```

Is there a smarter way to tell the build system that it should include it automatically? How do we do it if we mix it up. I would like to have it like this:

```
USEMODULE += pm_fallback_set_lowest
USEMODULE += pm_stm32_common_off
USEMODULE += pm_cortexm_common_reboot
```

but for now we would have to write the following:

```
USEMODULE += pm_fallback
USEMODULE += pm_fallback_set_lowest

USEMODULE += pm_stm32_common
USEMODULE += pm_stm32_common_off

USEMODULE += pm_cortexm_common
USEMODULE += pm_cortexm_common_reboot
```

**Comment by kaspar030** [comment]
> Is there a smarter way to tell the build system that it should include it automatically?

Ah, I forgot: add "pm_fallback" as dependency of "pm_falback_%" (e.g., in Makefile.dep):

```
ifneq (,$(filter pm_fallback_%,$(USEMODULE)))
  USEMODULE += pm_fallback
endif
```

**Comment by roberthartung** [comment]
> add "pm_fallback" as dependency of "pm_falback_%" (e.g., in Makefile.dep):

Where do add it? In each board, cpu, RIOT's Makefile.dep?

**Comment by roberthartung** [comment]
Additionally: is there a smart way to select all submodues automatically? e.g. pm_fallback will select _off, _set_lowest and _reboot?

**Comment by kaspar030** [comment]
> Where do add it? In each board, cpu, RIOT's Makefile.dep?

If the module is in drivers, drivers/Makefile.dep, otherwise ./Makefile.dep. The module is shared by all (or most) boards, right?

> Additionally: is there a smart way to select all submodues automatically? e.g. pm_fallback will select _off, _set_lowest and _reboot?

nope. ;)

**Comment by roberthartung** [comment]
Almost finished. One thing we really have to look at before merging: Can all CPUs that are based on cortexm_common by default use their pm implementation? That would be great!

**Comment by roberthartung** [comment]
@kaspar030 Pushed a submodule-based version. You can have a look now!

File: cpu/atmega1281/Makefile.include [hunk 1]
```diff
@@ -3,6 +3,9 @@ export CFLAGS += -DCOREIF_NG=1
 
 # tell the build system that the CPU depends on the atmega common files
 USEMODULE += atmega_common
+USEMODULE += pm_fallback_set_lowest
```
Comment (kaspar030) [inline cpu/atmega1281/Makefile.include]: can these be moved to atmega_common?

File: cpu/atmega_common/pm/pm.c [hunk 2]
```diff
@@ -26,6 +26,14 @@
 #include "irq.h"
 #include "periph/pm.h"
 
+void pm_set_lowest(void) {}
```
Comment (kaspar030) [inline cpu/atmega_common/pm/pm.c]: shouldn't this be in the common pm_fallback_set_lowest?

File: cpu/atmega_common/pm/pm.c [hunk 3]
```diff
@@ -26,6 +26,14 @@
...
+
+void pm_off(void)
```
Comment (kaspar030) [inline cpu/atmega_common/pm/pm.c]: dito

File: cpu/cortexm_common/Makefile.include [hunk 4]
```diff
@@ -3,5 +3,6 @@ INCLUDES += -I$(RIOTCPU)/cortexm_common/include
 INCLUDES += -I$(RIOTCPU)/cortexm_common/include/vendor
 
 USEMODULE += cortexm_common_periph
+include $(RIOTCPU)/cortexm_common/Makefile.dep
```
Comment (kaspar030) [inline cpu/cortexm_common/Makefile.include]: cannot include Makefile.dep from a different makefile

File: cpu/ezr32wg/Makefile.include [hunk 5]
```diff
@@ -1,3 +1,7 @@
 export CPU_ARCH = cortex-m4f
 
+USEMODULE += pm_cortexm_common_set_lowest
+USEMODULE += pm_cortexm_common_off
```
Comment (kaspar030) [inline cpu/ezr32wg/Makefile.include]: don't all cortexm  boards use the cortexm_common off and reboot?

File: cpu/lm4f120/Makefile.include [hunk 6]
```diff
@@ -1,8 +1,10 @@
 export CPU_ARCH = cortex-m4f
 
-include $(RIOTMAKE)/arch/cortexm.inc.mk
-
 # use common periph functions
 USEMODULE += periph_common
+USEMODULE += pm_cortexm_common_set_lowest
```
Comment (kaspar030) [inline cpu/lm4f120/Makefile.include]: IMO we can generalize this for all cortexm boards:
if not (feature_provided(periph/pm)) -> use ...._common_set_lowest
of make it dependend on whether pm_layered is used.

File: cpu/nrf5x_common/pm/pm.c [hunk 7]
```diff
@@ -23,6 +23,14 @@
 
 #include "cpu.h"
 
+void pm_set_lowest(void) {
```
Comment (kaspar030) [inline cpu/nrf5x_common/pm/pm.c]: use pm_fallback_coretexm_common_

File: drivers/pm_fallback/set_lowest.c [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ * @}
+ */
+
+#include "irq.h"
+#include "periph/pm.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void pm_set_lowest(void) {}
```
Comment (kaspar030) [inline drivers/pm_fallback/set_lowest.c]: I don't like this file much... ;)

File: cpu/atmega1281/Makefile.include (hunk 1 above)
Comment (roberthartung) [reply cpu/atmega1281/Makefile.include]: @kaspar030 No, the idea is that the *final* CPU will *always* select everything it needs. In particular my implementation of the pm is specific to atmega1284p, as registers and modes and peripheral levels are individual per CPU.
The only thing would be to create a pm_fallback.mk.inc in atmega_common that will be included in Makefile.include of the CPU?

File: cpu/atmega_common/pm/pm.c (hunk 3 above)
Comment (roberthartung) [reply cpu/atmega_common/pm/pm.c]: ACK

File: cpu/atmega_common/pm/pm.c (hunk 2 above)
Comment (roberthartung) [reply cpu/atmega_common/pm/pm.c]: ACK

File: cpu/cortexm_common/Makefile.include (hunk 4 above)
Comment (roberthartung) [reply cpu/cortexm_common/Makefile.include]: I think it didn't work without, but I will check again

File: cpu/lm4f120/Makefile.include (hunk 6 above)
Comment (roberthartung) [reply cpu/lm4f120/Makefile.include]: Sounds good. As mentioned above we have to check if all boards based on cortexm_common AND stm32_common can still use the pm implementation of cortexm_common!

File: cpu/ezr32wg/Makefile.include (hunk 5 above)
Comment (roberthartung) [reply cpu/ezr32wg/Makefile.include]: I think so

File: drivers/pm_fallback/set_lowest.c (hunk 8 above)
Comment (roberthartung) [reply drivers/pm_fallback/set_lowest.c]: Me neither, but that's the fallback and it was before as well. Any suggestions?

Comment (kaspar030) [reply drivers/pm_fallback/set_lowest.c]: Not with using submodules and without using ifdefs. Let's keep it for now.

File: cpu/lm4f120/Makefile.include (hunk 6 above)
Comment (kaspar030) [reply cpu/lm4f120/Makefile.include]: well, ```pm_reboot()``` will definitely be the same for all cortexm. "pm_off()" will be cpu dependent. "pm_set_lowest()" will be the same for all cortexm not implementing periph/pm.

File: drivers/pm_fallback/set_lowest.c (hunk 8 above)
Comment (roberthartung) [reply drivers/pm_fallback/set_lowest.c]: pm_set_lowest is only used in `core/kernel_init.c` and `sys/pm_layered/pm.c`, maybe we can omit the file and only call pm_set_lowest in `kernel_init.c` if the CPU/board defines something like `#define HAS_PM_SET_LOWEST` or something?

File: cpu/lm4f120/Makefile.include (hunk 6 above)
Comment (roberthartung) [reply cpu/lm4f120/Makefile.include]: where would we implemented such a check to select all cortexm_common fallback if the feature is not provided by the cpu itself?

**Comment by roberthartung** [comment]
From what I can see, all boards that are based on stm32_common, are also based on cortexm_common. Therefore, as stm32_common provides a pm_layered implementation, all boards based on stm32_common can select the pm_set method from stm32_common and off/reboot from cortexm_common.

```bash
find . -name "Makefile.include" | xargs grep -A2 stm32_common
```

**Comment by roberthartung** [comment]
@kaspar030 Can `kinetis_common` just use the default cortem_common pm implementation? You defined a single power mode here, which will in the end do the same as pm_set_lowest of pm_layered will call pm_set which calls cortexm_sleep(0), but in the end, this is the same as pm_set_lowest from cortexm_common.

**Comment by gebart** [comment]
Kinetis has 4 power modes which can be used with RIOT without any big intrusive changes. I have a WIP branch for adding them, but I have not yet posted a PR. Please leave the Kinetis specific implementation of pm_set in place to make merging my future PR easier, don't remove it right now.

**Comment by roberthartung** [comment]
@gebart Thanks! The function has been moved to its own module for now to comply with the other implementations. Might change that depending on the feedback receivedhere.
@kaspar030 I just committed all the changes requested.

**Comment by roberthartung** [comment]
I also updated the wiki page on the new power management API and the ideas behind it: https://github.com/RIOT-OS/RIOT/wiki/Power-Management#architecture

**Comment by cladmi** [comment]
I would prefer the version with submodules, because, the required code and the dependencies are explicit.

For me, the problem here, is the current possibility to express dependencies to interfaces which is a build system issue and not a source file issue.

**Comment by kaspar030** [comment]
Closed in favor of #7726.

@roberthartung Again, thanks a lot for the work you've put in this!

**Pull request closed without merging**